static/app/components/integrationExternalMappingForm.tsx

```diff
@@ -4,127 +4,153 @@ import capitalize from 'lodash/capitalize';
+  dataEndpoint: string;
+  onResults?: (data: any, mappingKey?: string) => void;
+  isInline?: boolean;
+  mappingKey?: string;
+};
 
 export default class IntegrationExternalMappingForm extends Component<Props> {
+  model = new FormModel();
+
   get initialData() {
     const {integration, mapping} = this.props;
-
     return {
-      externalName: '',
-      userId: '',
-      teamId: '',
-      sentryName: '',
       provider: integration.provider.key,
       integrationId: integration.id,
-      ...pick(mapping, ['externalName', 'userId', 'sentryName', 'teamId']),
+      ...pick(mapping, ['externalName', 'sentryName', 'userId', 'teamId']),
     };
   }
 
   get formFields(): Field[] {
-    const {type, sentryNamesMapper, url, mapping} = this.props;
+    const {
+      dataEndpoint,
+      isInline,
+      mapping,
+      mappingKey,
+      onResults,
+      sentryNamesMapper,
+      type,
+    } = this.props;
     const optionMapper = sentryNames =>
       sentryNames.map(({name, id}) => ({value: id, label: name}));
-
-    const fields: any[] = [
+    const fields: Field[] = [
       {
-        name: 'externalName',
-        type: 'string',
-        required: true,
-        label: tct('External [type]', {type: capitalize(type)}),
-        placeholder: t(`${type === 'team' ? '@org/teamname' : '@username'}`),
-      },
-    ];
-    if (type === 'user') {
-      fields.push({
-        name: 'userId',
+        name: `${type}Id`,
         type: 'select_async',
         required: true,
-        label: tct('Sentry [type]', {type: capitalize(type)}),
-        placeholder: t(`Choose your Sentry User`),
-        url,
+        label: isInline ? undefined : tct('Sentry [type]', {type: capitalize(type)}),
+        placeholder: t(`Select Sentry ${capitalize(type)}`),
+        url: dataEndpoint,
         onResults: result => {
-          // For organizations with >100 users, we want to make sure their
+          onResults?.(result, isInline ? mapping?.externalName : mappingKey);
+          // TODO(Leander): The code below only fixes the problem when viewed, not when edited
+          // Pagination still has bugs for results not on initial return of the query
+
+          // For organizations with >100 entries, we want to make sure their
           // saved mapping gets populated in the results if it wouldn't have
           // been in the initial 100 API results, which is why we add it here
-          if (mapping && !result.find(({user}) => user.id === mapping.userId)) {
-            result = [{id: mapping.userId, name: mapping.sentryName}, ...result];
+          if (
+            mapping &&
+            !result.find(entry => {
+              const id = type === 'user' ? entry.user.id : entry.id;
+              return id === mapping[`${type}Id`];
+            })
+          ) {
+            return optionMapper([
+              {id: mapping[`${type}Id`], name: mapping.sentryName},
+              ...sentryNamesMapper(result),
```

Comment:
The `sentryNamesMapper` maps the API returned object, to an object in the shape `{id: str, name: str}`. If we're pre-populating the field with a user/team for edits, we need to add it in manually if its NOT in the `dataEndpoint`'s first set of paginated results.

That is what the comment above is trying to solve.

Still though, it's a bit buggy with trying to make changes or typing-but-not-saving on the form. This is an issue with SelectAsyncField though, and not directly with this component.